File: src/world_services.py
#!/usr/bin/env python3

import math
import os
import random
from collections import deque

import rclpy
import numpy as np
from rclpy.node import Node
from scipy.spatial.transform import Rotation as R

from geometry_msgs.msg import PoseArray
from std_msgs.msg import Int64MultiArray

from override_sim.msg import ToggleArray, ToggleState
from override_sim.srv import FlipToggle, IntakeElement, LoadElement, ScoreElement

from ros_gz_interfaces.srv import DeleteEntity, SetEntityPose, SpawnEntity
from ros_gz_interfaces.msg import Entity, EntityFactory

from ament_index_python.packages import get_package_share_directory

# color codes: 1 = red, 2 = blue, 3 = yellow
PIN_MODELS = {
    (1, 3): 'pin-ry',
    (2, 3): 'pin-by',
    (3, 3): 'pin-yy',
    (1, 2): 'pin-rb',
}

# official element counts
ELEMENTS_LEFT_INIT = [20, 20, 19, 4, 56]   # ry, by, yy, rb, cups

GOAL_HEIGHTS = {
    0: 0.2227, 1: 0.1465, 2: 0.1465, 3: 0.0825, 4: 0.0825,
    5: 0.0825, 6: 0.0825, 7: 0.1465, 8: 0.1465,
}
GOAL_POSITIONS = {
    0: (0.0, 0.0), 1: (-0.6, 1.2), 2: (-1.2, 0.6), 3: (-1.2, -0.6),
    4: (-0.6, -1.2), 5: (0.6, 1.2), 6: (1.2, 0.6), 7: (1.2, -0.6),
    8: (0.6, -1.2),
}
LOADER_POSITIONS = {
    0: (-1.74, 1.49), 1: (-1.74, -1.49), 2: (1.74, 1.49), 3: (1.74, -1.49),
}
# toggle world poses: position + wall yaw; state roll = state * 120 deg
TOGGLE_POSES = {
    0: (0.0, 1.78, 0.0),
    1: (1.78, 0.0, math.pi / 2),
    2: (0.0, -1.78, 0.0),
    3: (-1.78, 0.0, math.pi / 2),
}


class WorldServices(Node):
    """Owns world-side game actions: robot intake, element placement on goals,
    match loading onto loaders, and Toggle flipping. Also publishes toggle
    state and the match phase."""

    def __init__(self):
        super().__init__('world_services')

        self.declare_parameter('world_name', 'override')
        self.declare_parameter('intake_capacity', 10)
        self.declare_parameter('autonomous', False)
        self.world_name = self.get_parameter('world_name').value
        self.intake_capacity = self.get_parameter('intake_capacity').value
        self.autonomous = self.get_parameter('autonomous').value

        self.pkg_path = get_package_share_directory('override_sim')

        # robot pose (for dropping misplaced elements)
        self.create_subscription(PoseArray, '/otto_pose', self.robot_pose_callback, 10)

        # publishers
        self.robot_elements = self.create_publisher(
            Int64MultiArray, '/robot_elements', 10)
        self.elements_remaining = self.create_publisher(
            Int64MultiArray, '/elements_remaining', 10)
        self.toggles_pub = self.create_publisher(ToggleArray, '/toggles', 10)
        self.phase_pub = self.create_publisher(Int64MultiArray, '/game_phase', 10)

        # services
        self.srv_intake = self.create_service(
            IntakeElement, '/robot_intake', self.intake_element)
        self.srv_score = self.create_service(
            ScoreElement, '/score_element', self.score_element)
        self.srv_loader = self.create_service(
            LoadElement, '/loader', self.load_element)
        self.srv_toggle = self.create_service(
            FlipToggle, '/flip_toggle', self.flip_toggle)

        # gazebo service clients
        self.remove_entity = self.create_client(
            DeleteEntity, f'/world/{self.world_name}/remove')
        self.spawn_entity = self.create_client(
            SpawnEntity, f'/world/{self.world_name}/create')
        self.set_pose = self.create_client(
            SetEntityPose, f'/world/{self.world_name}/set_pose')

        # state
        self.robot_intake = deque(maxlen=self.intake_capacity)
        self.goal_stack_count = {gid: 0 for gid in GOAL_HEIGHTS}
        self.toggle_states = {tid: 0 for tid in TOGGLE_POSES}
        self.elements_left = list(ELEMENTS_LEFT_INIT)
        self.name_counters = {}

        self.publish_elements_remaining()
        self.publish_toggles()

        # match phase management
        self.phase = 1
        if self.autonomous:
            self.phase = 0
            self.create_timer(15.0, self.end_autonomous)
        self.create_timer(1.0, self.publish_phase)

    # ------------------------------------------------------------------
    # callbacks
    # ------------------------------------------------------------------
    def robot_pose_callback(self, msg: PoseArray):
        self.robot_x = msg.poses[-1].position.x
        self.robot_y = msg.poses[-1].position.y

    def end_autonomous(self):
        if self.phase == 0:
            self.get_logger().info("autonomous period ended")
            self.phase = 1

    def publish_phase(self):
        msg = Int64MultiArray()
        msg.data = [self.phase]
        self.phase_pub.publish(msg)

    # ------------------------------------------------------------------
    # services
    # ------------------------------------------------------------------
    def intake_element(self, request: IntakeElement.Request, response):
        if len(self.robot_intake) >= self.intake_capacity:
            self.get_logger().info("robot intake full")
            response.success = False
            return response

        self.delete_entity(request.entity_id)
        self.robot_intake.append(
            (request.entity_type, request.top_color, request.bottom_color))
        self.get_logger().info(
            f"robot intake: {list(self.robot_intake)}")
        self.publish_robot_elements()
        response.success = True
        return response

    def score_element(self, request: ScoreElement.Request, response):
        # pick the element to place: explicit colors win, else first in intake
        if request.top_color != 0 or request.bottom_color != 0:
            el_type = request.element_type
            top, bottom = request.top_color, request.bottom_color
        else:
            if not self.robot_intake:
                self.get_logger().info("nothing to place")
                response.success = False
                return response
            el_type, top, bottom = self.robot_intake.popleft()

        goal_id = request.goal_id
        if goal_id not in GOAL_POSITIONS:
            # not at a goal: drop near the robot
            x = self.robot_x + random.uniform(-0.3, 0.3)
            y = self.robot_y + random.uniform(-0.3, 0.3)
            self.spawn_element(el_type, top, bottom, x, y, 0.3)
            self.get_logger().info("dropped element near robot")
        else:
            gx, gy = GOAL_POSITIONS[goal_id]
            z = GOAL_HEIGHTS[goal_id] + self.goal_stack_count[goal_id] * 0.17 + 0.02
            self.spawn_element(el_type, top, bottom, gx, gy, z)
            self.goal_stack_count[goal_id] += 1
            self.get_logger().info(
                f"placed element type {el_type} on goal {goal_id}")

        self.publish_robot_elements()
        response.success = True
        return response

    def load_element(self, request: LoadElement.Request, response):
        loader_id = request.loader_id
        if loader_id not in LOADER_POSITIONS:
            response.success = False
            return response

        model_name = self.element_model(
            request.element_type, request.top_color, request.bottom_color)
        if model_name is None or not self.take_one_element(
                request.element_type, request.top_color, request.bottom_color):
            self.get_logger().info("no matching elements left")
            response.success = False
            return response

        lx, ly = LOADER_POSITIONS[loader_id]
        self.spawn_element(request.element_type, request.top_color,
                           request.bottom_color, lx, ly, 0.45)
        self.get_logger().info(f"loaded element onto loader {loader_id}")
        self.publish_elements_remaining()
        response.success = True
        return response

    def flip_toggle(self, request: FlipToggle.Request, response):
        toggle_id = request.toggle_id
        if toggle_id not in TOGGLE_POSES:
            response.success = False
            return response

        if request.state < 0:
            new_state = (self.toggle_states[toggle_id] + 1) % 3
        else:
            new_state = request.state % 3

        self.toggle_states[toggle_id] = new_state
        self.rotate_toggle(toggle_id, new_state)
        self.publish_toggles()
        self.get_logger().info(
            f"toggle {toggle_id} -> state {new_state}")
        response.success = True
        response.new_state = new_state
        return response

    # ------------------------------------------------------------------
    # gazebo helpers
    # ------------------------------------------------------------------
    def element_model(self, el_type, top, bottom):
        if el_type == 1:                      # pin
            return PIN_MODELS.get((bottom, top))
        if el_type == 2:                      # cup
            return 'cup'
        return None

    def spawn_element(self, el_type, top, bottom, x, y, z, yaw=0.0):
        model = self.element_model(el_type, top, bottom)
        if model is None:
            self.get_logger().warning("unknown element, not spawning")
            return

        counter = self.name_counters.setdefault(model, 100)
        self.name_counters[model] = counter + 1
        name = f"{model.replace('-', '_')}_{counter}"

        factory = EntityFactory()
        factory.name = name
        factory.allow_renaming = True
        factory.sdf_filename = os.path.join(
            self.pkg_path, 'models', model, 'model.sdf')
        factory.pose.position.x = x
        factory.pose.position.y = y
        factory.pose.position.z = z
        q = R.from_euler('z', yaw).as_quat()
        factory.pose.orientation.x = q[0]
        factory.pose.orientation.y = q[1]
        factory.pose.orientation.z = q[2]
        factory.pose.orientation.w = q[3]

        req = SpawnEntity.Request()
        req.entity_factory = factory
        self.spawn_entity.call_async(req)

    def delete_entity(self, entity_id):
        entity = Entity()
        entity.id = int(entity_id)
        req = DeleteEntity.Request()
        req.entity = entity
        self.remove_entity.call_async(req)

    def rotate_toggle(self, toggle_id, state):
        x, y, base_yaw = TOGGLE_POSES[toggle_id]
        roll = state * (2 * math.pi / 3)
        q = (R.from_euler('z', base_yaw) * R.from_euler('x', roll)).as_quat()
        req = SetEntityPose.Request()
        req.entity.name = f'toggle_{toggle_id + 1}'
        req.pose.position.x = x
        req.pose.position.y = y
        req.pose.position.z = 0.0
        req.pose.orientation.x = q[0]
        req.pose.orientation.y = q[1]
        req.pose.orientation.z = q[2]
        req.pose.orientation.w = q[3]
        self.set_pose.call_async(req)

    # ------------------------------------------------------------------
    # bookkeeping publishers
    # ------------------------------------------------------------------
    def take_one_element(self, el_type, top, bottom):
        idx = 4 if el_type == 2 else list(PIN_MODELS).index((bottom, top))
        if self.elements_left[idx] <= 0:
            return False
        self.elements_left[idx] -= 1
        return True

    def publish_robot_elements(self):
        msg = Int64MultiArray()
        msg.data = []
        for el_type, top, bottom in self.robot_intake:
            msg.data.extend([el_type, top, bottom])
        self.robot_elements.publish(msg)

    def publish_elements_remaining(self):
        msg = Int64MultiArray()
        msg.data = self.elements_left
        self.elements_remaining.publish(msg)

    def publish_toggles(self):
        msg = ToggleArray()
        quadrant_map = {0: 'N', 1: 'E', 2: 'S', 3: 'W'}
        for tid in sorted(self.toggle_states):
            t = ToggleState()
            t.toggle_id = tid
            t.state = self.toggle_states[tid]
            t.quadrant = quadrant_map[tid]
            msg.toggles.append(t)
        self.toggles_pub.publish(msg)


def main(args=None):
    rclpy.init(args=args)
    node = WorldServices()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/world_services.cpp
// world_services.cpp - owns world-side game actions: robot intake, element
// placement on goals, match loading onto loaders, and Toggle flipping.
// Also publishes toggle state and the match phase.
#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdint>
#include <deque>
#include <map>
#include <memory>
#include <random>
#include <string>
#include <utility>
#include <vector>

#include <ament_index_cpp/get_package_share_directory.hpp>
#include <geometry_msgs/msg/pose_array.hpp>
#include <rclcpp/rclcpp.hpp>
#include <ros_gz_interfaces/msg/entity.hpp>
#include <ros_gz_interfaces/msg/entity_factory.hpp>
#include <ros_gz_interfaces/srv/delete_entity.hpp>
#include <ros_gz_interfaces/srv/set_entity_pose.hpp>
#include <ros_gz_interfaces/srv/spawn_entity.hpp>
#include <std_msgs/msg/int64_multi_array.hpp>

#include "override_sim/field_constants.h"
#include "override_sim/msg/toggle_array.hpp"
#include "override_sim/msg/toggle_state.hpp"
#include "override_sim/srv/flip_toggle.hpp"
#include "override_sim/srv/intake_element.hpp"
#include "override_sim/srv/load_element.hpp"
#include "override_sim/srv/score_element.hpp"

namespace override_sim
{

using DeleteEntity = ros_gz_interfaces::srv::DeleteEntity;
using SpawnEntity = ros_gz_interfaces::srv::SpawnEntity;
using SetEntityPose = ros_gz_interfaces::srv::SetEntityPose;

class WorldServices : public rclcpp::Node
{
public:
  WorldServices() : Node("world_services")
  {
    this->declare_parameter<std::string>("world_name", "override");
    this->declare_parameter<int>("intake_capacity", 10);
    this->declare_parameter<bool>("autonomous", false);
    this->get_parameter("world_name", world_name_);
    this->get_parameter("intake_capacity", intake_capacity_);
    this->get_parameter("autonomous", autonomous_);

    pkg_path_ = ament_index_cpp::get_package_share_directory("override_sim");

    robot_pose_sub_ = this->create_subscription<geometry_msgs::msg::PoseArray>(
      "/otto_pose", 10,
      [this](const geometry_msgs::msg::PoseArray::SharedPtr msg) {
        if (!msg->poses.empty()) {
          robot_x_ = msg->poses.back().position.x;
          robot_y_ = msg->poses.back().position.y;
        }
      });

    robot_elements_pub_ = this->create_publisher<std_msgs::msg::Int64MultiArray>(
      "/robot_elements", 10);
    elements_remaining_pub_ =
      this->create_publisher<std_msgs::msg::Int64MultiArray>(
      "/elements_remaining", 10);
    toggles_pub_ = this->create_publisher<ToggleArray>("/toggles", 10);
    phase_pub_ = this->create_publisher<std_msgs::msg::Int64MultiArray>(
      "/game_phase", 10);

    intake_srv_ = this->create_service<IntakeElement>("/robot_intake",
      [this](const std::shared_ptr<IntakeElement::Request> req,
        std::shared_ptr<IntakeElement::Response> res) { handle_intake(req, res); });
    score_srv_ = this->create_service<ScoreElement>("/score_element",
      [this](const std::shared_ptr<ScoreElement::Request> req,
        std::shared_ptr<ScoreElement::Response> res) { handle_score(req, res); });
    loader_srv_ = this->create_service<LoadElement>("/loader",
      [this](const std::shared_ptr<LoadElement::Request> req,
        std::shared_ptr<LoadElement::Response> res) { handle_load(req, res); });
    toggle_srv_ = this->create_service<FlipToggle>("/flip_toggle",
      [this](const std::shared_ptr<FlipToggle::Request> req,
        std::shared_ptr<FlipToggle::Response> res) { handle_flip(req, res); });

    remove_entity_ = this->create_client<DeleteEntity>(
      "/world/" + world_name_ + "/remove");
    spawn_entity_ = this->create_client<SpawnEntity>(
      "/world/" + world_name_ + "/create");
    set_pose_ = this->create_client<SetEntityPose>(
      "/world/" + world_name_ + "/set_pose");

    for (const auto & kv : kGoals) {
      goal_stack_count_[kv.first] = 0;
    }
    for (const auto & kv : kToggles) {
      toggle_states_[kv.first] = 0;
    }
    elements_left_ = {20, 20, 19, 4, 56};

    publish_elements_remaining();
    publish_toggles();

    phase_ = 1;
    if (autonomous_) {
      phase_ = 0;
      auton_timer_ = this->create_wall_timer(
        std::chrono::seconds(15), [this]() { end_autonomous(); });
    }
    phase_timer_ = this->create_wall_timer(
      std::chrono::seconds(1), [this]() { publish_phase(); });
  }

private:
  // ------------------------------------------------------------------
  // phase
  // ------------------------------------------------------------------
  void end_autonomous()
  {
    if (phase_ == 0) {
      RCLCPP_INFO(this->get_logger(), "autonomous period ended");
      phase_ = 1;
    }
  }

  void publish_phase()
  {
    auto msg = std::make_unique<std_msgs::msg::Int64MultiArray>();
    msg->data.push_back(phase_);
    phase_pub_->publish(std::move(msg));
  }

  // ------------------------------------------------------------------
  // services
  // ------------------------------------------------------------------
  void handle_intake(
    const std::shared_ptr<IntakeElement::Request> req,
    std::shared_ptr<IntakeElement::Response> res)
  {
    if (static_cast<int>(robot_intake_.size()) >= intake_capacity_) {
      RCLCPP_INFO(this->get_logger(), "robot intake full");
      res->success = false;
      return;
    }
    delete_entity(req->entity_id);
    robot_intake_.push_back(
      {req->entity_type, req->top_color, req->bottom_color});
    RCLCPP_INFO(this->get_logger(), "robot intake: size %zu", robot_intake_.size());
    publish_robot_elements();
    res->success = true;
  }

  void handle_score(
    const std::shared_ptr<ScoreElement::Request> req,
    std::shared_ptr<ScoreElement::Response> res)
  {
    int8_t el_type = req->element_type;
    int8_t top = req->top_color;
    int8_t bottom = req->bottom_color;

    if (top == 0 && bottom == 0) {
      if (robot_intake_.empty()) {
        RCLCPP_INFO(this->get_logger(), "nothing to place");
        res->success = false;
        return;
      }
      auto e = robot_intake_.front();
      robot_intake_.pop_front();
      el_type = e[0];
      top = e[1];
      bottom = e[2];
    }

    const auto goal_it = kGoals.find(req->goal_id);
    if (goal_it == kGoals.end()) {
      // not at a goal: drop near the robot
      std::uniform_real_distribution<double> dist(-0.3, 0.3);
      const double x = robot_x_ + dist(rng_);
      const double y = robot_y_ + dist(rng_);
      spawn_element(el_type, top, bottom, x, y, 0.3);
      RCLCPP_INFO(this->get_logger(), "dropped element near robot");
    } else {
      const GoalInfo & g = goal_it->second;
      const double z = g.height +
        goal_stack_count_[req->goal_id] * 0.17 + 0.02;
      spawn_element(el_type, top, bottom, g.x, g.y, z);
      goal_stack_count_[req->goal_id] += 1;
      RCLCPP_INFO(this->get_logger(), "placed element type %d on goal %d",
        el_type, req->goal_id);
    }

    publish_robot_elements();
    res->success = true;
  }

  void handle_load(
    const std::shared_ptr<LoadElement::Request> req,
    std::shared_ptr<LoadElement::Response> res)
  {
    const auto loader_it = kLoaders.find(req->loader_id);
    if (loader_it == kLoaders.end()) {
      res->success = false;
      return;
    }
    const std::string model = element_model(
      req->element_type, req->top_color, req->bottom_color);
    if (model.empty() || !take_one_element(
        req->element_type, req->top_color, req->bottom_color))
    {
      RCLCPP_INFO(this->get_logger(), "no matching elements left");
      res->success = false;
      return;
    }
    spawn_element(req->element_type, req->top_color, req->bottom_color,
      loader_it->second.first, loader_it->second.second, 0.45);
    RCLCPP_INFO(this->get_logger(), "loaded element onto loader %d",
      req->loader_id);
    publish_elements_remaining();
    res->success = true;
  }

  void handle_flip(
    const std::shared_ptr<FlipToggle::Request> req,
    std::shared_ptr<FlipToggle::Response> res)
  {
    const auto toggle_it = kToggles.find(req->toggle_id);
    if (toggle_it == kToggles.end()) {
      res->success = false;
      return;
    }
    int new_state;
    if (req->state < 0) {
      new_state = (toggle_states_[req->toggle_id] + 1) % 3;
    } else {
      new_state = req->state % 3;
    }
    toggle_states_[req->toggle_id] = static_cast<int8_t>(new_state);
    rotate_toggle(req->toggle_id, new_state);
    publish_toggles();
    RCLCPP_INFO(this->get_logger(), "toggle %d -> state %d",
      req->toggle_id, new_state);
    res->success = true;
    res->new_state = static_cast<int8_t>(new_state);
  }

  // ------------------------------------------------------------------
  // gazebo helpers
  // ------------------------------------------------------------------
  static std::string element_model(int8_t el_type, int8_t top, int8_t bottom)
  {
    if (el_type == 1) {  // pin
      switch ((bottom << 8) | top) {
        case (1 << 8) | 3: return "pin-ry";
        case (2 << 8) | 3: return "pin-by";
        case (3 << 8) | 3: return "pin-yy";
        case (1 << 8) | 2: return "pin-rb";
        default: return "";
      }
    }
    if (el_type == 2) {  // cup
      return "cup";
    }
    return "";
  }

  void spawn_element(
    int8_t el_type, int8_t top, int8_t bottom,
    double x, double y, double z, double yaw = 0.0)
  {
    const std::string model = element_model(el_type, top, bottom);
    if (model.empty()) {
      RCLCPP_WARN(this->get_logger(), "unknown element, not spawning");
      return;
    }

    int counter = 100;
    const auto it = name_counters_.find(model);
    if (it != name_counters_.end()) {
      counter = it->second;
    }
    name_counters_[model] = counter + 1;
    std::string model_flat = model;
    std::replace(model_flat.begin(), model_flat.end(), '-', '_');
    const std::string name = model_flat + "_" + std::to_string(counter);

    auto factory = std::make_shared<ros_gz_interfaces::msg::EntityFactory>();
    factory->name = name;
    factory->allow_renaming = true;
    factory->sdf_filename = pkg_path_ + "/models/" + model + "/model.sdf";
    factory->pose.position.x = x;
    factory->pose.position.y = y;
    factory->pose.position.z = z;
    double qx, qy, qz, qw;
    QuatFromYaw(yaw, &qx, &qy, &qz, &qw);
    factory->pose.orientation.x = qx;
    factory->pose.orientation.y = qy;
    factory->pose.orientation.z = qz;
    factory->pose.orientation.w = qw;

    auto req = std::make_shared<SpawnEntity::Request>();
    req->entity_factory = *factory;
    if (!spawn_entity_->service_is_ready()) {
      RCLCPP_WARN(this->get_logger(), "spawn service not ready");
      return;
    }
    spawn_entity_->async_send_request(req);
  }

  void delete_entity(int32_t entity_id)
  {
    auto req = std::make_shared<DeleteEntity::Request>();
    req->entity.id = static_cast<uint64_t>(entity_id);
    if (!remove_entity_->service_is_ready()) {
      RCLCPP_WARN(this->get_logger(), "remove service not ready");
      return;
    }
    remove_entity_->async_send_request(req);
  }

  void rotate_toggle(int toggle_id, int state)
  {
    const auto & t = kToggles.at(toggle_id);
    const double roll = state * (2.0 * M_PI / 3.0);
    double qz_x, qz_y, qz_z, qz_w;
    QuatFromYaw(t[2], &qz_x, &qz_y, &qz_z, &qz_w);
    const double qx_x = std::sin(roll / 2.0);
    const double qx_w = std::cos(roll / 2.0);
    double qx0, qy0, qz0, qw0;
    QuatMul(qz_x, qz_y, qz_z, qz_w, qx_x, 0.0, 0.0, qx_w,
      &qx0, &qy0, &qz0, &qw0);

    auto req = std::make_shared<SetEntityPose::Request>();
    req->entity.name = "toggle_" + std::to_string(toggle_id + 1);
    req->pose.position.x = t[0];
    req->pose.position.y = t[1];
    req->pose.position.z = 0.0;
    req->pose.orientation.x = qx0;
    req->pose.orientation.y = qy0;
    req->pose.orientation.z = qz0;
    req->pose.orientation.w = qw0;
    if (!set_pose_->service_is_ready()) {
      RCLCPP_WARN(this->get_logger(), "set_pose service not ready");
      return;
    }
    set_pose_->async_send_request(req);
  }

  // ------------------------------------------------------------------
  // bookkeeping publishers
  // ------------------------------------------------------------------
  bool take_one_element(int8_t el_type, int8_t top, int8_t bottom)
  {
    int idx;
    if (el_type == 2) {
      idx = 4;
    } else {
      const std::string model = element_model(el_type, top, bottom);
      if (model == "pin-ry") idx = 0;
      else if (model == "pin-by") idx = 1;
      else if (model == "pin-yy") idx = 2;
      else if (model == "pin-rb") idx = 3;
      else return false;
    }
    if (elements_left_[idx] <= 0) {
      return false;
    }
    elements_left_[idx] -= 1;
    return true;
  }

  void publish_robot_elements()
  {
    auto msg = std::make_unique<std_msgs::msg::Int64MultiArray>();
    for (const auto & e : robot_intake_) {
      msg->data.push_back(e[0]);
      msg->data.push_back(e[1]);
      msg->data.push_back(e[2]);
    }
    robot_elements_pub_->publish(std::move(msg));
  }

  void publish_elements_remaining()
  {
    auto msg = std::make_unique<std_msgs::msg::Int64MultiArray>();
    for (int v : elements_left_) {
      msg->data.push_back(v);
    }
    elements_remaining_pub_->publish(std::move(msg));
  }

  void publish_toggles()
  {
    auto msg = std::make_unique<ToggleArray>();
    for (const auto & kv : kToggles) {
      ToggleState t;
      t.toggle_id = kv.first;
      t.state = toggle_states_.at(kv.first);
      t.quadrant = kQuadrantByToggle[kv.first];
      msg->toggles.push_back(t);
    }
    toggles_pub_->publish(std::move(msg));
  }

  // state
  std::string world_name_;
  int intake_capacity_ = 10;
  bool autonomous_ = false;
  std::string pkg_path_;
  double robot_x_ = 0.0;
  double robot_y_ = 0.0;
  std::deque<std::array<int8_t, 3>> robot_intake_;
  std::map<int, int> goal_stack_count_;
  std::map<int, int8_t> toggle_states_;
  std::array<int, 5> elements_left_;
  std::map<std::string, int> name_counters_;
  std::mt19937 rng_{std::random_device{}()};
  int phase_ = 1;

  rclcpp::Subscription<geometry_msgs::msg::PoseArray>::SharedPtr robot_pose_sub_;
  rclcpp::Publisher<std_msgs::msg::Int64MultiArray>::SharedPtr
    robot_elements_pub_;
  rclcpp::Publisher<std_msgs::msg::Int64MultiArray>::SharedPtr
    elements_remaining_pub_;
  rclcpp::Publisher<ToggleArray>::SharedPtr toggles_pub_;
  rclcpp::Publisher<std_msgs::msg::Int64MultiArray>::SharedPtr phase_pub_;
  rclcpp::Service<IntakeElement>::SharedPtr intake_srv_;
  rclcpp::Service<ScoreElement>::SharedPtr score_srv_;
  rclcpp::Service<LoadElement>::SharedPtr loader_srv_;
  rclcpp::Service<FlipToggle>::SharedPtr toggle_srv_;
  rclcpp::Client<DeleteEntity>::SharedPtr remove_entity_;
  rclcpp::Client<SpawnEntity>::SharedPtr spawn_entity_;
  rclcpp::Client<SetEntityPose>::SharedPtr set_pose_;
  rclcpp::TimerBase::SharedPtr auton_timer_;
  rclcpp::TimerBase::SharedPtr phase_timer_;
};

}  // namespace override_sim

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<override_sim::WorldServices>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
